cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/SSHSessionManager.java

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.distributed.remote;
+
+import co.cask.cdap.common.ssh.DefaultSSHSession;
+import co.cask.cdap.common.ssh.SSHConfig;
+import co.cask.cdap.internal.app.runtime.monitor.SSHSessionProvider;
+import co.cask.cdap.internal.app.runtime.monitor.proxy.MonitorSocksProxy;
+import co.cask.cdap.runtime.spi.ssh.PortForwarding;
+import co.cask.cdap.runtime.spi.ssh.SSHProcess;
+import co.cask.cdap.runtime.spi.ssh.SSHSession;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.InetSocketAddress;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.annotation.Nullable;
+
+/**
+ * Manages SSH sessions for the runtime {@link MonitorSocksProxy}.
+ */
+final class SSHSessionManager implements SSHSessionProvider, AutoCloseable {
+
+  private final ConcurrentMap<String, SSHInfo> sshInfos;
```

Comment:
Isn't this problematic for the Remote Hadoop Provisioner, as the hostname won't be unique across program runs?

Reply:
hum, the removal could be a problem. Need to have RunId as part of the key.

Reply:
The provisioner and runtime itself should be ok for concurrent runs, as they have different run if, hence different directory and runtime monitoring keys

Reply:
I updated the second commit with the correct implementation. Please take a look again.